security/tpm/src/tpm.ts

```diff
@@ -3,22 +3,423 @@
+   */
+  hierarchy: TPM_HANDLE;
+
+  /**
+   * Handle value where the persistent key is expected to be found
+   */
+  handle: TPM_HANDLE;
+
+  /**
+   * Template to be used for key cretaion if the persistent key with the given handle does not exist
+   */
+  template: TPMT_PUBLIC;
+}
+
 
 export class TpmSecurityClient  {
-  getEndorsementKey(callback: (err?: Error) => void): void {
-    throw new errors.NotImplementedError();
+
+  private _ek: TPMT_PUBLIC = null;
+  private _srk: TPMT_PUBLIC = null;
+  private _registrationId: string = '';
+  private _tpm: Tpm;
+  private _fsm: machina.Fsm;
+  private _idKeyPub: TPMT_PUBLIC = null;
+
+
+  constructor(registrationId?: string, customTpm?: any) {
+    this._tpm = customTpm ? customTpm : new Tpm(false);
+    if (registrationId) {
+      this._registrationId = registrationId;
+    }
+    this._fsm = new machina.Fsm({
+      initialState: 'disconnected',
+      states: {
+        disconnected: {
+          _onEnter: (callback, err) => {
+            this._ek = null;
+            this._srk = null;
+            if (callback) {
+              if (err) {
+                callback(err);
+              } else {
+                callback(null, null);
+              }
+            }
+          },
+          connect: (connectCallback) => this._fsm.transition('connecting', connectCallback),
+          getEndorsementKey: (callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('getEndorsementKey', callback);
+              }
+            });
+          },
+          getStorageRootKey: (callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('getStorageRootKey', callback);
+              }
+            });
+          },
+          signWithIdentity: (dataToSign, callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('signWithIdentity', dataToSign, callback);
+              }
+            });
+          },
+          activateSymmetricIdentity: (identityKey, callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('activateSymmetricIdentity', identityKey, callback);
+              }
+            });
+          },
+          disconnect: (callback) => {
+            if (callback) {
+              callback();
+            }
+           }
+        },
+        connecting: {
+          _onEnter: (callback) => {
+            try {
+              let self = this;
+              this._tpm.connect(function (): void {
```

Comment:
>function (): void [](start = 32, length = 17)

interesting syntax.  Is there a reason you chose this instead of the typical "() => {" form?

Reply:
Fixed.

---
In reply to: [159289053](https://github.com/Azure/azure-iot-sdk-node/pull/185#discussion_r159289053) [](ancestors = 159289053)